### ISSUE TITLE
vault agent with templates doesn't work with `missing client token`

### ISSUE THREAD

**Issue by gitirabassi** [hashicorp/vault#7790]
```
ov 05 13:45:48 etcd vault[2657]: ==> Vault server started! Log data will stream in below:
Nov 05 13:45:48 etcd vault[2657]: ==> Vault agent configuration:
Nov 05 13:45:48 etcd vault[2657]:            Api Address 1: unix:///var/run/vault_agent.sock
Nov 05 13:45:48 etcd vault[2657]:                      Cgo: disabled
Nov 05 13:45:48 etcd vault[2657]:                Log Level: info
Nov 05 13:45:48 etcd vault[2657]:                  Version: Vault v1.3.0-beta1
Nov 05 13:45:48 etcd vault[2657]: 2019-11-05T13:45:48.247Z [INFO]  sink.file: creating file sink
Nov 05 13:45:48 etcd vault[2657]: 2019-11-05T13:45:48.247Z [INFO]  sink.file: file sink configured: path=/tmp/agent_token mode=-rw-r-----
Nov 05 13:45:48 etcd vault[2657]: 2019-11-05T13:45:48.249Z [INFO]  auth.handler: starting auth handler
Nov 05 13:45:48 etcd vault[2657]: 2019-11-05T13:45:48.249Z [INFO]  auth.handler: authenticating
Nov 05 13:45:48 etcd vault[2657]: 2019-11-05T13:45:48.250Z [INFO]  template.server: starting template server
Nov 05 13:45:48 etcd vault[2657]: [INFO] (runner) creating new runner (dry: false, once: false)
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (runner) final config: {"Consul":{"Address":"","Auth":{"Enabled":false,"Username":"","Password":""},"Retry":{"Attempts":12,"Backoff":250000000,"MaxBackoff":60000000000,"Enabled":true},"SSL":{"CaCert":"","CaPath":"","Cert":"","Enabled":false,"Key":"","ServerName":"","Verify":true},"Token":"","Transport":{"DialKeepAlive":30000000000,"DialTimeout":30000000000,"DisableKeepAlives":false,"IdleConnTimeout":90000000000,"MaxIdleConns":100,"MaxIdleConnsPerHost":3,"TLSHandshakeTimeout":10000000000}},"Dedup":{"Enabled":false,"MaxStale":2000000000,"Prefix":"consul-template/dedup/","TTL":15000000000},"Exec":{"Command":"","Enabled":false,"Env":{"Blacklist":[],"Custom":[],"Pristine":false,"Whitelist":[]},"KillSignal":2,"KillTimeout":30000000000,"ReloadSignal":null,"Splay":0,"Timeout":0},"KillSignal":2,"LogLevel":"WARN","MaxStale":2000000000,"PidFile":"","ReloadSignal":1,"Syslog":{"Enabled":false,"Facility":"LOCAL0"},"Templates":[{"Backup":false,"Command":"","CommandTimeout":30000000000,"Contents":"","CreateDestDirs":true,"Destination":"/etc/etcd/pki/server.key","ErrMissingKey":false,"Exec":{"Command":"","Enabled":false,"Env":{"Blacklist":[],"Custom":[],"Pristine":false,"Whitelist":[]},"KillSignal":2,"KillTimeout":30000000000,"ReloadSignal":null,"Splay":0,"Timeout":30000000000},"Perms":0,"Source":"/etc/vault/etcd-server-key.hcl","Wait":{"Enabled":false,"Min":0,"Max":0},"LeftDelim":"","RightDelim":"","FunctionBlacklist":null,"SandboxPath":""},{"Backup":false,"Command":"","CommandTimeout":30000000000,"Contents":"","CreateDestDirs":true,"Destination":"/etc/etcd/pki/server.crt","ErrMissingKey":false,"Exec":{"Command":"","Enabled":false,"Env":{"Blacklist":[],"Custom":[],"Pristine":false,"Whitelist":[]},"KillSignal":2,"KillTimeout":30000000000,"ReloadSignal":null,"Splay":0,"Timeout":30000000000},"Perms":0,"Source":"/etc/vault/etcd-server-certificate.hcl","Wait":{"Enabled":false,"Min":0,"Max":0},"LeftDelim":"","RightDelim":"","FunctionBlacklist":null,"SandboxPath":""}],"Vault":{"Address":"http://192.168.33.10:8200","Enabled":true,"Grace":15000000000,"Namespace":"","RenewToken":false,"Retry":{"Attempts":12,"Backoff":250000000,"MaxBackoff":60000000000,"Enabled":true},"SSL":{"CaCert":"","CaPath":"","Cert":"","Enabled":false,"Key":"","ServerName":"","Verify":false},"Transport":{"DialKeepAlive":30000000000,"DialTimeout":30000000000,"DisableKeepAlives":false,"IdleConnTimeout":90000000000,"MaxIdleConns":100,"MaxIdleConnsPerHost":3,"TLSHandshakeTimeout":10000000000},"UnwrapToken":false},"Wait":{"Enabled":false,"Min":0,"Max":0},"Once":false}
Nov 05 13:45:48 etcd vault[2657]: [INFO] (runner) creating watcher
Nov 05 13:45:48 etcd vault[2657]: 2019-11-05T13:45:48.251Z [INFO]  sink.server: starting sink server
Nov 05 13:45:48 etcd vault[2657]: 2019-11-05T13:45:48.256Z [INFO]  auth.handler: authentication successful, sending token to sinks
Nov 05 13:45:48 etcd vault[2657]: 2019-11-05T13:45:48.256Z [INFO]  auth.handler: starting renewal process
Nov 05 13:45:48 etcd vault[2657]: 2019-11-05T13:45:48.256Z [INFO]  sink.file: token written: path=/tmp/agent_token
Nov 05 13:45:48 etcd vault[2657]: 2019-11-05T13:45:48.257Z [INFO]  template.server: template server received new token
Nov 05 13:45:48 etcd vault[2657]: [INFO] (runner) stopping
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (runner) stopping watcher
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (watcher) stopping all views
Nov 05 13:45:48 etcd vault[2657]: [INFO] (runner) creating new runner (dry: false, once: false)
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (runner) final config: {"Consul":{"Address":"","Auth":{"Enabled":false,"Username":"","Password":""},"Retry":{"Attempts":12,"Backoff":250000000,"MaxBackoff":60000000000,"Enabled":true},"SSL":{"CaCert":"","CaPath":"","Cert":"","Enabled":false,"Key":"","ServerName":"","Verify":true},"Token":"","Transport":{"DialKeepAlive":30000000000,"DialTimeout":30000000000,"DisableKeepAlives":false,"IdleConnTimeout":90000000000,"MaxIdleConns":100,"MaxIdleConnsPerHost":3,"TLSHandshakeTimeout":10000000000}},"Dedup":{"Enabled":false,"MaxStale":2000000000,"Prefix":"consul-template/dedup/","TTL":15000000000},"Exec":{"Command":"","Enabled":false,"Env":{"Blacklist":[],"Custom":[],"Pristine":false,"Whitelist":[]},"KillSignal":2,"KillTimeout":30000000000,"ReloadSignal":null,"Splay":0,"Timeout":0},"KillSignal":2,"LogLevel":"WARN","MaxStale":2000000000,"PidFile":"","ReloadSignal":1,"Syslog":{"Enabled":false,"Facility":"LOCAL0"},"Templates":[{"Backup":false,"Command":"","CommandTimeout":30000000000,"Contents":"","CreateDestDirs":true,"Destination":"/etc/etcd/pki/server.key","ErrMissingKey":false,"Exec":{"Command":"","Enabled":false,"Env":{"Blacklist":[],"Custom":[],"Pristine":false,"Whitelist":[]},"KillSignal":2,"KillTimeout":30000000000,"ReloadSignal":null,"Splay":0,"Timeout":30000000000},"Perms":0,"Source":"/etc/vault/etcd-server-key.hcl","Wait":{"Enabled":false,"Min":0,"Max":0},"LeftDelim":"","RightDelim":"","FunctionBlacklist":null,"SandboxPath":""},{"Backup":false,"Command":"","CommandTimeout":30000000000,"Contents":"","CreateDestDirs":true,"Destination":"/etc/etcd/pki/server.crt","ErrMissingKey":false,"Exec":{"Command":"","Enabled":false,"Env":{"Blacklist":[],"Custom":[],"Pristine":false,"Whitelist":[]},"KillSignal":2,"KillTimeout":30000000000,"ReloadSignal":null,"Splay":0,"Timeout":30000000000},"Perms":0,"Source":"/etc/vault/etcd-server-certificate.hcl","Wait":{"Enabled":false,"Min":0,"Max":0},"LeftDelim":"","RightDelim":"","FunctionBlacklist":null,"SandboxPath":""}],"Vault":{"Address":"http://192.168.33.10:8200","Enabled":true,"Grace":15000000000,"Namespace":"","RenewToken":false,"Retry":{"Attempts":12,"Backoff":250000000,"MaxBackoff":60000000000,"Enabled":true},"SSL":{"CaCert":"","CaPath":"","Cert":"","Enabled":false,"Key":"","ServerName":"","Verify":false},"Transport":{"DialKeepAlive":30000000000,"DialTimeout":30000000000,"DisableKeepAlives":false,"IdleConnTimeout":90000000000,"MaxIdleConns":100,"MaxIdleConnsPerHost":3,"TLSHandshakeTimeout":10000000000},"UnwrapToken":false},"Wait":{"Enabled":false,"Min":0,"Max":0},"Once":false}
Nov 05 13:45:48 etcd vault[2657]: [INFO] (runner) creating watcher
Nov 05 13:45:48 etcd vault[2657]: [INFO] (runner) starting
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (runner) running initial templates
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (runner) initiating run
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (runner) checking template f58cadff8b7bf41783f25a67de2d7de8
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (runner) was not watching 1 dependencies
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (watcher) adding vault.write(cluster/staging/etcd/issue/server -> 4b8e4995)
Nov 05 13:45:48 etcd vault[2657]: [TRACE] (watcher) vault.write(cluster/staging/etcd/issue/server -> 4b8e4995) starting
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (runner) checking template 8586dbe4c1aa15b543fd2964d0a2ffee
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (runner) missing data for 1 dependencies
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (runner) diffing and updating dependencies
Nov 05 13:45:48 etcd vault[2657]: [DEBUG] (runner) watching 1 dependencies
Nov 05 13:45:48 etcd vault[2657]: [TRACE] (view) vault.write(cluster/staging/etcd/issue/server -> 4b8e4995) starting fetch
Nov 05 13:45:48 etcd vault[2657]: [TRACE] vault.write(cluster/staging/etcd/issue/server -> 4b8e4995): PUT /v1/cluster/staging/etcd/issue/server?stale=true&wait=1m0s
Nov 05 13:45:48 etcd vault[2657]: 2019-11-05T13:45:48.259Z [INFO]  auth.handler: renewed auth token
Nov 05 13:45:51 etcd vault[2657]: [WARN] (view) vault.write(cluster/staging/etcd/issue/server -> 4b8e4995): vault.write(cluster/staging/etcd/issue/server -> 4b8e4995): Error making API request.
                                  
                                  URL: PUT http://192.168.33.10:8200/v1/cluster/staging/etcd/issue/server
                                  Code: 400. Errors:
                                  
                                  * missing client token (retry attempt 1 after "250ms")
Nov 05 13:45:52 etcd vault[2657]: [TRACE] (view) vault.write(cluster/staging/etcd/issue/server -> 4b8e4995) starting fetch
Nov 05 13:45:52 etcd vault[2657]: [TRACE] vault.write(cluster/staging/etcd/issue/server -> 4b8e4995): PUT /v1/cluster/staging/etcd/issue/server?stale=true&wait=1m0s
Nov 05 13:45:56 etcd vault[2657]: [WARN] (view) vault.write(cluster/staging/etcd/issue/server -> 4b8e4995): vault.write(cluster/staging/etcd/issue/server -> 4b8e4995): Error making API request.
                                  
                                  URL: PUT http://192.168.33.10:8200/v1/cluster/staging/etcd/issue/server
                                  Code: 400. Errors:
                                  
                                  * missing client token (retry attempt 2 after "500ms")
Nov 05 13:45:56 etcd vault[2657]: [TRACE] (view) vault.write(cluster/staging/etcd/issue/server -> 4b8e4995) starting fetch
Nov 05 13:45:56 etcd vault[2657]: [TRACE] vault.write(cluster/staging/etcd/issue/server -> 4b8e4995): PUT /v1/cluster/staging/etcd/issue/server?stale=true&wait=1m0s
Nov 05 13:46:00 etcd vault[2657]: [WARN] (view) vault.write(cluster/staging/etcd/issue/server -> 4b8e4995): vault.write(cluster/staging/etcd/issue/server -> 4b8e4995): Error making API request.
```

Also vault agent doesn't exit with `CTRL+C` when templating is on

**Comment by gitirabassi** [comment]
probably related to https://github.com/hashicorp/vault/pull/7739

**Comment by catsby** [comment]
Hello - thanks for the logs. Can you share more about the template you're trying to render? I'm curious about the error shown here and it's relation to templating:

```
Nov 05 13:45:56 etcd vault[2657]: [WARN] (view) vault.write(cluster/staging/etcd/issue/server -> 4b8e4995): vault.write(cluster/staging/etcd/issue/server -> 4b8e4995): Error making API request.
                                  
                                  URL: PUT http://192.168.33.10:8200/v1/cluster/staging/etcd/issue/server
                                  Code: 400. Errors:
                                  
                                  * missing client token (retry attempt 2 after "500ms")
```

The Agent uses the token generated from the auto-auth, so it's possible that the token lacks permissions? Seeing more of the configuration would likely help, if you can do so without exposing any secret information.

Thanks!

**Comment by gitirabassi** [comment]
At first this was the case, but now I have it configured with `consul-template` directly and it works perfectly

**Comment by gitirabassi** [comment]
I’ll try again after that PR gets merged. Do you have an eta on merging that

**Comment by gitirabassi** [comment]
Just verified and this was fixed with `1.3.0` release

**Comment by catsby** [comment]
Thanks for the follow-up!
